src/log.rs

```diff
@@ -1,47 +1,152 @@
-use crate::include::common::validate::validate_input;
-use crate::include::dav1d::dav1d::Rav1dLogger;
-use crate::src::internal::Rav1dContext;
-use crate::stderr;
+use cstr::cstr;
 use std::ffi::c_char;
-use std::ffi::c_int;
+use std::ffi::c_uint;
 use std::ffi::c_void;
+use std::fmt;
+use std::fmt::Write as _;
+use std::io::stderr;
+use std::io::stdout;
+use std::io::Write as _;
 use std::ptr;
 
-extern "C" {
-    fn vfprintf(_: *mut libc::FILE, _: *const c_char, _: ::core::ffi::VaList) -> c_int;
+pub type Dav1dLoggerCallback = unsafe extern "C" fn(
+    // The above `cookie` field.
+    cookie: *mut c_void,
+    // A `printf`-style format specifier.
+    fmt: *const c_char,
+    // `printf`-style variadic args.
+    args: ...
+);
+
+#[derive(Clone)]
+#[repr(C)]
+pub struct Dav1dLogger {
+    /// A cookie that's passed as the first argument to the callback below.
+    cookie: *mut c_void,
+    /// A `printf`-style function except for an extra first argument that will always be the above `cookie`.
+    callback: Option<Dav1dLoggerCallback>,
 }
 
-#[cold]
-pub unsafe extern "C" fn rav1d_log_default_callback(
-    _cookie: *mut c_void,
-    format: *const c_char,
-    mut ap: ::core::ffi::VaList,
-) {
-    vfprintf(stderr, format, ap.as_va_list());
+impl Dav1dLogger {
+    /// # Safety
+    ///
+    /// `callback`, if non-[`None`]/`NULL` must be safe to call when:
+    /// * the first argument is `cookie`
+    /// * the rest of the arguments would be safe to call `printf` with
+    pub const unsafe fn new(cookie: *mut c_void, callback: Option<Dav1dLoggerCallback>) -> Self {
+        Self { cookie, callback }
+    }
 }
 
-impl Default for Rav1dLogger {
+impl Default for Dav1dLogger {
     fn default() -> Self {
-        Self {
-            cookie: ptr::null_mut(),
-            callback: Some(rav1d_log_default_callback),
+        // Safety: `callback` is [`None`].
+        unsafe { Self::new(ptr::null_mut(), None) }
+    }
```

Comment:
This changes the default logging functionality to not log, rather than log to stderr?

Reply:
Hmm, I may have messed up here.  Let me look at where this default is used and I'll make sure it stays stderr.

Reply:
Actually, I'll remove this `impl Default` entirely.  It's only used by the binaries and they can stay is )`Dav1dLogger { cookie: 0 as *mut c_void, callback: None }`).

Reply:
Removed in 881dafe.